File: api/src/main/java/net/blueshell/api/event/EventEventListener.java
package net.blueshell.api.event;

import net.blueshell.api.common.event.PostRemoveEvent;
import net.blueshell.api.common.event.PostUpdateEvent;
import net.blueshell.api.common.event.PrePersistEvent;
import net.blueshell.api.job.calendar.AddEventToCalendarJob;
import net.blueshell.api.job.calendar.RemoveEventFromCalendarJob;
import net.blueshell.api.job.calendar.SyncEventToCalendarJob;
import net.blueshell.api.model.event.Event;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

@Component
public class EventEventListener {

    private final AddEventToCalendarJob addJob;
    private final SyncEventToCalendarJob syncJob;
    private final RemoveEventFromCalendarJob removeJob;

    public EventEventListener(AddEventToCalendarJob addJob,
                              SyncEventToCalendarJob syncJob,
                              RemoveEventFromCalendarJob removeJob) {
        this.addJob = addJob;
        this.syncJob = syncJob;
        this.removeJob = removeJob;
    }

    /**
     * After commit, enqueue add if approved
     */
    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void onPersist(PrePersistEvent<Event> evt) {
        Event e = evt.getSource();
        if (e.isApproved()) {
            addJob.add(e.getId());
        }
    }

    /**
     * After commit, enqueue sync if approved, otherwise remove
     */
    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void onUpdate(PostUpdateEvent<Event> evt) {
        Event e = evt.getSource();
        if (e.isApproved()) {
            syncJob.sync(e.getId());
        } else {
            removeJob.remove(e.getId());
        }
    }

    /**
     * After commit, enqueue remove
     */
    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void onDelete(PostRemoveEvent<Event> evt) {
        Event e = evt.getSource();
        removeJob.remove(e.getId());
    }
}

File: api/src/main/kotlin/net/blueshell/api/event/application/listener/EventEventListener.kt
package net.blueshell.api.event.application.listener

import net.blueshell.api.event.application.EventService
import net.blueshell.api.event.application.event.EventChangeType
import net.blueshell.api.event.application.event.EventChangedEvent
import net.blueshell.api.platform.integration.event.job.AddEventToCalendarEvent
import net.blueshell.api.platform.integration.event.job.RemoveEventFromCalendarEvent
import net.blueshell.api.platform.integration.event.job.SyncEventToCalendarEvent
import org.springframework.context.ApplicationEventPublisher
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional

@Component
class EventEventListener(
    private val eventPublisher: ApplicationEventPublisher,
    private val events: EventService
) {
    /**
     * After commit, enqueue add if approved
     */
    @EventListener
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    fun onChange(evt: EventChangedEvent) {
        when (evt.changeType) {
            EventChangeType.CREATED -> {
                val e = events.findById(evt.eventId)
                if (e.approved) {
                    eventPublisher.publishEvent(AddEventToCalendarEvent(e.id))
                }
            }
            EventChangeType.UPDATED -> {
                val e = events.findById(evt.eventId)
                if (e.approved) {
                    eventPublisher.publishEvent(SyncEventToCalendarEvent(e.id))
                } else {
                    eventPublisher.publishEvent(RemoveEventFromCalendarEvent(e.id))
                }
            }
            EventChangeType.DELETED -> eventPublisher.publishEvent(RemoveEventFromCalendarEvent(evt.eventId))
        }
    }
}
